File: Algorithms/quicksort.cpp
//QUICKSORT
#include <iostream>
using namespace std;

void swap(int* a, int* b)
{
	int temp = *a;
	*a = *b;
	*b = temp;
}

int partition(int arr[], int start, int end)
{
	int pivot = arr[end];
	int i = start-1;
	for(int j = start; j < end; j++)
	{
		if(arr[j] <= pivot)
		{
			i++;
			swap(&arr[j], &arr[i]);
		}
	}
	swap(arr[i+1], arr[end]);
	return i+1;
}

void quicksort(int arr[], int start, int end)
{
	if(start < end)
	{
		int splitpoint = partition(arr, start, end);
		quicksort(arr, start, splitpoint-1);
		quicksort(arr, splitpoint+1, end);
	}
}

int main() 
{
	int arr[] = {7, 6, 2, 1, 0, 9, 4, 8};
	quicksort(arr, 0, 7);
	for(int i = 0; i < 8; i++)
	    cout<<arr[i]<<" ";
	return 0;
}

File: coding-practice/Sort/quick_sort.py
def quicksort(arr, low, high):
    if low < high:
        pi = partition(arr, low, high)
        quicksort(arr, low, pi-1)
        quicksort(arr, pi+1, high)


#....pivot....

def partition(arr, l, h):
    i = l-1
    pivot_index = h
    pivot = arr[pivot_index]
    for j in range(l, pivot_index):
        if arr[j] < pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]

    arr[i+1], arr[pivot_index] = arr[pivot_index], arr[i+1]
    return i+1

arr = [10, 7, 8, -9, 1, 5]
quicksort(arr, 0, 5)
print(arr)